diff -aruN 2.3.3-01-asan-fix/src/sysc/kernel/sc_cor_qt.cpp 2.3.3-02-valgrind/src/sysc/kernel/sc_cor_qt.cpp
--- 2.3.3-01-asan-fix/src/sysc/kernel/sc_cor_qt.cpp	2021-07-28 10:46:22.241280298 +0200
+++ 2.3.3-02-valgrind/src/sysc/kernel/sc_cor_qt.cpp	2021-07-28 11:17:28.398980253 +0200
@@ -175,6 +175,9 @@
     cor->m_stack_size = stack_size;
     cor->m_stack = mmap(NULL, cor->m_stack_size, PROT_READ | PROT_WRITE,
                         MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+#ifdef HAVE_VALGRIND_H
+    cor->m_vgid = VALGRIND_STACK_REGISTER(cor->m_stack, (char*)cor->m_stack + cor->m_stack_size - 1);
+#endif
     void* sto = stack_align( cor->m_stack, QUICKTHREADS_STKALIGN,
                              &cor->m_stack_size );
     cor->m_sp = QUICKTHREADS_SP(sto, cor->m_stack_size - QUICKTHREADS_STKALIGN);
diff -aruN 2.3.3-01-asan-fix/src/sysc/kernel/sc_cor_qt.h 2.3.3-02-valgrind/src/sysc/kernel/sc_cor_qt.h
--- 2.3.3-01-asan-fix/src/sysc/kernel/sc_cor_qt.h	2021-07-28 11:00:10.428070067 +0200
+++ 2.3.3-02-valgrind/src/sysc/kernel/sc_cor_qt.h	2021-07-28 11:51:30.424547262 +0200
@@ -35,6 +35,10 @@
 
 #include <sys/mman.h> // for munmap
 
+#ifdef HAVE_VALGRIND_H
+#include <valgrind/valgrind.h>
+#endif
+
 #include "sysc/kernel/sc_cor.h"
 #include "sysc/packages/qt/qt.h"
 
@@ -57,11 +61,18 @@
     // constructor
     sc_cor_qt()
 	: m_stack_size( 0 ), m_stack( 0 ), m_sp( 0 ), m_pkg( 0 )
+#ifdef HAVE_VALGRIND_H
+	, m_vgid( 0 )
+#endif
 	{}
 
     // destructor
-    virtual ~sc_cor_qt()
-        { munmap(m_stack, m_stack_size); }
+    virtual ~sc_cor_qt() {
+#ifdef HAVE_VALGRIND_H
+        VALGRIND_STACK_DEREGISTER(m_vgid);
+#endif
+        munmap(m_stack, m_stack_size);
+    }
 
     // switch stack protection on/off
     virtual void stack_protect( bool enable );
@@ -74,6 +85,10 @@
 
     sc_cor_pkg_qt* m_pkg;         // the creating coroutine package
 
+#ifdef HAVE_VALGRIND_H
+    unsigned int   m_vgid;        // valgrind stack id
+#endif
+
 private:
 
     // disabled
