diff -aruN 2.3.3-01-asan-fix/src/sysc/kernel/sc_cor_qt.cpp 2.3.3-02-valgrind/src/sysc/kernel/sc_cor_qt.cpp
--- 2.3.3-01-asan-fix/src/sysc/kernel/sc_cor_qt.cpp	2021-07-29 19:54:05.265259317 +0200
+++ 2.3.3-02-valgrind/src/sysc/kernel/sc_cor_qt.cpp	2021-07-29 19:55:54.101255480 +0200
@@ -196,6 +196,9 @@
     cor->m_stack_size = stack_size;
     cor->m_stack = mmap(NULL, cor->m_stack_size, PROT_READ | PROT_WRITE,
                         MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+#ifdef HAVE_VALGRIND_H
+    cor->m_vgid = VALGRIND_STACK_REGISTER(cor->m_stack, (char*)cor->m_stack + cor->m_stack_size - 1);
+#endif
     void* sto = stack_align( cor->m_stack, QUICKTHREADS_STKALIGN,
                              &cor->m_stack_size );
     cor->m_sp = QUICKTHREADS_SP(sto, cor->m_stack_size - QUICKTHREADS_STKALIGN);
diff -aruN 2.3.3-01-asan-fix/src/sysc/kernel/sc_cor_qt.h 2.3.3-02-valgrind/src/sysc/kernel/sc_cor_qt.h
--- 2.3.3-01-asan-fix/src/sysc/kernel/sc_cor_qt.h	2021-07-29 19:55:16.669260145 +0200
+++ 2.3.3-02-valgrind/src/sysc/kernel/sc_cor_qt.h	2021-07-29 19:55:23.953259215 +0200
@@ -35,6 +35,10 @@
 
 #include <sys/mman.h> // for munmap
 
+#ifdef HAVE_VALGRIND_H
+#include <valgrind/valgrind.h>
+#endif
+
 #include "sysc/kernel/sc_cor.h"
 #include "sysc/packages/qt/qt.h"
 
@@ -57,10 +61,16 @@
     // constructor
     sc_cor_qt()
 	: m_stack_size( 0 ), m_stack( 0 ), m_sp( 0 ), m_pkg( 0 )
+#ifdef HAVE_VALGRIND_H
+        , m_vgid( 0 )
+#endif
 	{}
 
     // destructor
     virtual ~sc_cor_qt() {
+#ifdef HAVE_VALGRIND_H
+        VALGRIND_STACK_DEREGISTER(m_vgid);
+#endif
         if (m_stack)
             munmap(m_stack, m_stack_size);
     }
@@ -76,6 +86,10 @@
 
     sc_cor_pkg_qt* m_pkg;         // the creating coroutine package
 
+#ifdef HAVE_VALGRIND_H
+    unsigned int   m_vgid;        // valgrind stack id
+#endif
+
 private:
 
     // disabled
