diff --git a/docs/CMakeLists.txt b/docs/CMakeLists.txt
index 5156b342..9296040a 100644
--- a/docs/CMakeLists.txt
+++ b/docs/CMakeLists.txt
@@ -48,7 +48,7 @@ foreach (DOC_FILE ${DOC_FILES})
   get_filename_component (DOC_INSTALL_PATH ${DOC_FILE} PATH)
   install (FILES ${DOC_FILE}
            DESTINATION ${CMAKE_INSTALL_DOCDIR}/${DOC_INSTALL_PATH}
-           COMPONENT doc)
+           COMPONENT doc OPTIONAL)
 endforeach (DOC_FILE ${DOC_FILES})
 
 if (BUILD_SOURCE_DOCUMENTATION)
